### PR TITLE
Add derive path

### PR DESCRIPTION
**Pull request by chesterliliang**
1. Add derive path
2. Upgrade wookong-solo to 0.1.4
3. Path not send to device due to new firmware needed. 

Plan to add new command 'select' for switch multi seed.
```
fn get_seed_amount() ->u32
fn select(public_key, path) -> Result<<Self::Pair as TraitPair>::Public>;
fn choose(index) -> Result<<Self::Pair as TraitPair>::Public>;
```

### CODE REVIEW

File: core/primitives/src/crypto.rs [hunk 1]
```diff
@@ -704,7 +704,7 @@ pub trait Pair: CryptoType + Sized + Clone + Send + Sync + 'static {
 				return Ok(r)
 			}
 		}
-
+
```
Comment (kianenigma) [inline core/primitives/src/crypto.rs]:
```suggestion
```

File: subkey/subhd/Cargo.toml [hunk 2]
```diff
@@ -7,7 +7,7 @@ edition = "2018"
 [dependencies]
 hex-view = "0.1.3"
 primitives = { package = "substrate-primitives",  version = "*", path = "../../core/primitives" }
-wookong-solo = "0.1.2"
+wookong-solo = {version = "0.1.4"}
```
Comment (kianenigma) [inline subkey/subhd/Cargo.toml]:
```suggestion
wookong-solo = { version = "0.1.4" }
```

File: subkey/subhd/src/lib.rs [hunk 3]
```diff
@@ -70,10 +74,16 @@ pub struct Wookong;
 
 impl Wallet for Wookong {
 	type Pair = sr25519::Pair;
-	fn derive_public(&self, _path: &[DeriveJunction]) -> Result<Public> {
-		// TODO: pass through _path and _password to mutate the key on the device accordingly.
+	fn derive_public(&self, path: &[DeriveJunction]) -> Result<Public> {
 		let mut pk: [u8; 32] = [0u8; 32];
-		let rv = wk_getpub(&mut pk);
+		let mut vec = Vec::new();
+		for j in path {
```
Comment (kianenigma) [inline subkey/subhd/src/lib.rs]: I am no expert in derivation paths nor the hardware wallet integration, but regardless I don't get this loop: 

you are matching to check the type of junction, but you are doing the same thing regardless of the result?

File: subkey/subhd/src/lib.rs [hunk 4]
```diff
@@ -35,6 +35,10 @@ pub enum Error {
 	DeviceNotInit,
 	/// General error of command.
 	DeviceError,
+	/// No seed matching the give account
+	NoAccount,
```
Comment (kianenigma) [inline subkey/subhd/src/lib.rs]: Are these used anywhere? or is this a WIP PR?

File: subkey/subhd/src/lib.rs (hunk 3 above)
Comment (gavofyork) [reply subkey/subhd/src/lib.rs]: indeed; this is definitely wrong behaviour. the device needs to derive correctly according to the path.

**Comment by kianenigma** [comment]
any updates?

**Comment by chesterliliang** [comment]
@kianenigma We are working on the firmware part https://github.com/wookong-dot/sr25519-embedded-c
As soon as it is finished, we will test it with subkey and may submit fixes.

**Comment by chesterliliang** [comment]
@burdges the implement of soft derive in schnorrkel calls 'witness_bytes' in 'derive_secret_key' which need rand_hack() to get random number for nonce. Could you please make it the 'attach_rng' to let extern random in to do that?

**Comment by burdges** [comment]
It's now fixed in master, thanks!  I created that problem when I removed the Clone bound on transcripts.  I fixed it by adding a `derive_key_simple_rng` method in https://github.com/w3f/schnorrkel/commit/ad2190b1c0bbb204690a9959f7cb9427cb073273

**Comment by chesterliliang** [comment]
@burdges Thanks a lot. It works well.

**Comment by bkchr** [comment]
Any update?

**Comment by chesterliliang** [comment]
We finished the derive work in the firmware but realized we need deliver another EVB to you for the test. That will not be a good way since the update will come up constantly. So we decide to find some way to simulate the behavior of firmware on desktop OS. QEMU is the first choice and I found some samples written in RUST, so I  can not hold myself to try to use RUST to write entire firmware.... Not a wise plan and wasted a lot of time. So may be you could close this PR and we will do it again when ready.

**Comment by bkchr** [comment]
Okay :)

**Pull request closed without merging**